Add support for xyz.amorgan.blurhash in ImageContent

redstrate opened the pull request:
This adds an optional key to ImageContent, for displaying a blurhash on the client before an image is loaded.

Example consumer in Neochat: https://invent.kde.org/network/neochat/-/merge_requests/1151

redstrate committed:
Add support for xyz.amorgan.blurhash in ImageContent

This adds an optional key to ImageContent, for displaying a blurhash on
the client before an image is loaded.

redstrate committed:
Add support for xyz.amorgan.blurhash in ImageContent

This adds an optional key to ImageContent, for displaying a blurhash on
the client before an image is loaded.

KitsuneRal commented:
I'd be happy to see the blurhash generation code in the library too - would you mind moving it over here from the NeoChat MR? :)

redstrate commented:
> I'd be happy to see the blurhash generation code in the library too - would you mind moving it over here from the NeoChat MR? :)

Sure can!

redstrate committed:
Add blurhash implementation to Quotient

So every client that wants to implement blurhashes doesn't have to copy
this file over and over.

redstrate committed:
Add blurhash implementation to Quotient

So every client that wants to implement blurhashes doesn't have to copy
this file over and over.

redstrate commented:
Okay the blurhash implementation has been moved here, will update the NeoChat MR accordingly

KitsuneRal commented on Quotient/blurhash.cpp:
```diff
@@ -0,0 +1,464 @@
+}
+}
+
+#ifdef DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
```
I'd appreciate it if you could move this piece out to `tests/` and port it from Doctest to QTest that we use there; or I can do it after merging, too, not a blocker.

redstrate replied:
Oops I forgot there were tests in this same file, I'll try porting them

redstrate replied:
Done

KitsuneRal commented on Quotient/blurhash.cpp:
```diff
@@ -0,0 +1,464 @@
+
+int encodeMaxAC(float maxAC) noexcept
+{
+    return std::max(0, std::min(82, int(maxAC * 166 - 0.5f)));
```
```suggestion
    return std::clamp(static_cast<int>(maxAC * 166 - 0.5f), 0, 82);
```

redstrate replied:
Done

KitsuneRal reviewed (changes requested):
Sorry for waiting; overall LGTM save for the tests

KitsuneRal commented on Quotient/blurhash.cpp:
```diff
@@ -0,0 +1,274 @@
+
+#include "blurhash.h"
+
+#include <QColorSpace>
```
```suggestion
#include <QtGui/QColorSpace>
```

redstrate replied:
Done

KitsuneRal commented on Quotient/blurhash.h:
```diff
@@ -0,0 +1,107 @@
+
+#include "quotient_export.h"
+
+#include <QImage>
```
```suggestion
#include <QtGui/QImage>
```

redstrate replied:
Done

KitsuneRal commented on Quotient/blurhash.h:
```diff
@@ -0,0 +1,107 @@
+     */
+    static QUOTIENT_API QString encode(const QImage &image, int componentsX = 4, int componentsY = 4);
+
+protected:
```
`protected` implies inheriting from a class - do you expect that clients should derive from this class and somehow use its static internals?

redstrate replied:
No that's just for the tests, hence why the only friend is `TestBlurHash`. I added a `final` to the class to make sure a client doesn't think to subclass it.

KitsuneRal commented on Quotient/blurhash.h:
```diff
@@ -0,0 +1,107 @@
+    static QUOTIENT_API QString encode(const QImage &image, int componentsX = 4, int componentsY = 4);
+
+protected:
+    struct Components {
```
Why not alias `QPoint` or `std::pair`?

redstrate replied:
Changed to a std::pair, good idea. I didn't go with a QPoint because it's not _technically_ a point, it just has X/Y fields.

KitsuneRal reviewed (commented):
Mostly LGTM but I have a question regarding the `protected` piece. I actually thought it would be merely two functions in the interface but you seem to intend something more, apparently?

KitsuneRal commented:
Okay I took a look at it once again and the same question from above holds: what's the point of having the protected interface and the whole class in fact? Why not a namespace with two functions and the rest pulled into the .cpp backend?

redstrate commented:
Because I over-engineered it initially :D I can change it to what you want

redstrate commented:
Another Q: should the blurhash generation be handled within Quotient itself or left up to the application? NeoChat would basically do something like this:

```cpp
    QImage image(url.toLocalFile());
    content = new EventContent::ImageContent(url, fileInfo.size(), mime, image.size(), fileInfo.fileName(), BlurHash::encode(image));
```

redstrate commented:
> Another Q: should the blurhash generation be handled within Quotient itself or left up to the application? NeoChat would basically do something like this:
> 
> ```c++
>     QImage image(url.toLocalFile());
>     content = new EventContent::ImageContent(url, fileInfo.size(), mime, image.size(), fileInfo.fileName(), BlurHash::encode(image));
> ```

Actually nvm, I misread our own code as we create a QImage anyway for grabbing the size. This is pretty much the same use-case.
